Add AutoRegistering protocol for automatic cell registration

apptekstudios opened the pull request:
We implemented auto-Registration of tableViewCell types. This was done such that it is an optional, non-breaking change. The features is disabled by default, although that could be changed without breaking anything.

When calling dequeueReusableCell(...), one can now specify autoRegister: true, and it will take care of ensuring that the class/nib is registered. Note that in order to get around generic constraints this required duplicating the dequeueReusableCell function for each of Reusable and NibReusable.

Hopefully this is useful to someone!

apptekstudios committed:
Implement ‘AutoRegistering’ Protocol
- Updated examples
- Updated docs

AliSoftware commented on Example/ReusableDemo iOS/TableViewController.swift:
```diff
@@ -14,9 +14,10 @@ final class TableViewController: UITableViewController {
-    tableView.register(cellType: MySimpleColorCell.self)
-    tableView.register(cellType: MyXIBTextCell.self)
-    tableView.register(cellType: MyXIBInfoCell.self)
+    /* Since we're using autoRegister there's no need to register these type */
```
Typo: s/type/types/

AliSoftware commented on Sources/View/UITableView+Reusable.swift:
```diff
@@ -40,25 +40,61 @@ public extension UITableView {
    - seealso: `dequeueReusableCell(withIdentifier:,for:)`
    */
-  final func dequeueReusableCell<T: UITableViewCell>(for indexPath: IndexPath, cellType: T.Type = T.self) -> T
+    final func dequeueReusableCell<T: UITableViewCell>(
```
Please use 2-spaces indentation for consistency's stake with the rest of the code (I'm not strongly an advocate for 2 vs 4 spaces, but since we use 2 everywhere else, better keep it consistent)

AliSoftware commented on Sources/View/UITableView+Reusable.swift:
```diff
@@ -40,25 +40,61 @@ public extension UITableView {
+            if let bareCell = self.dequeueReusableCell(withIdentifier: cellType.reuseIdentifier) as? T {
+                return bareCell //Get a cell if we can
+            } else {
+                register(cellType: cellType) //Register class or nib
```
I know I'm a little OCD but please add a space after the `//` at the very start of the comment (here and on line 58 just above)

AliSoftware commented on Example/ReusableDemo iOS/TableViewController.swift:
```diff
@@ -14,9 +14,10 @@ final class TableViewController: UITableViewController {
+    /* Since we're using autoRegister there's no need to register these type */
+    //tableView.register(cellType: MySimpleColorCell.self)
+    //tableView.register(cellType: MyXIBTextCell.self)
+    //tableView.register(cellType: MyXIBInfoCell.self)
```
For the sake of example and to show how to register cells manually with Reusable, maybe only auto-register one of the 3? Or keep one of the 3 manual?

AliSoftware commented on Sources/View/UITableView+Reusable.swift:
```diff
@@ -40,25 +40,61 @@ public extension UITableView {
       return cell
   }
 
+    final func dequeueReusableCell<T: UITableViewCell>(
```
Same remark about indentation

AliSoftware reviewed (changes requested):
Some nitpickings.

Also, I'm not sure about that generic constraints limitation you talked about that forced you to duplicate the code, can you elaborate about the error you get if you don't duplicate?

apptekstudios commented:
The generic constraints issue relates to the `dequeueReusableCell<T: UITableViewCell>` function casting T as 'Reusable', meaning that the subsequent call to `register(cellType:)` tries to register the class rather than the nib (in cases where it is a NibReusable).

Swift doesn't seem to like conditionally casting to a protocol, so this was the simple workaround. There might be a non-duplicating way around this that I haven't thought of though!

AliSoftware reviewed (commented):
Looking good!

I think maybe we could still find a way to factorize the two same methods (which differ only from their `where` clause) to make them call a common private method.

After all the only difference is that each variant of this method calls a different variant of `register(cellType:)`.

So if for example we provide a private helper method which takes a `register: (() -> Void)?` closure as a parameter (and no `autoRegister: Bool`), and do all the common stuff common to these two methods in there, then our two variants of the real public functions will just need to call that private helper function with `registerClosure: autoRegister ? { register(cellType: cellType) } : nil` — avoiding to duplicate the common code. Even if we'd still need to duplicate the methods, at least the implementations would be factorized that way

What do you think?

AliSoftware commented:
FYI the CI failure is due to the following linter errors:

```
…
Linting 'MyCustomWidget.swift' (25/25)
/Users/distiller/project/Example/ReusableDemo iOS/TableViewController.swift:19: warning: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
/Users/distiller/project/Example/ReusableDemo iOS/TableViewController.swift:66: warning: Line Length Violation: Line should be 120 characters or less: currently 146 characters (line_length)
```

AliSoftware commented:
@apptekstudios I took the liberty to push a commit on your fork to show my idea of mutualising the code. (Also added missing documentation on the new method)

There are still things missing in that PR imho:
* You implemented this for `UITableView.dequeueReusableCell`, but why not for all the other stuff? Like `dequeueHeaderFooter`, and like for the similar methods in `UICollectionViews` as well?
* The more I think about this, the more I wonder if it wouldn't be nicer — both for API and for usage — to use additional protocol conformance instead of adding the `autoRegister: Bool` parameter (that you'll have to not forget to add at call site on every place you dequeue your cells otherwise)?

So maybe it would be more powerful, flexible, and less error-prone for the end user, to declare a `protocol AutoRegistering {}` for example in Reusable, and make different implementations for `where T: NibReusable, T: AutoRegistering` vs just `where T: NibReusable` (the compiler always choose the most specific implementation anyway)?

That way users of the Reusable library would just have to just add `AutoRegistering` to their custom `UITableViewCell` subclasses (the same way they just add `Reusable` or `NibReusable` or `NibLoadable` already), making auto-registration still an opt-in feature, but without having to think about explicitly adding `autoRegister: true` everywhere they want to dequeue such cells. This would also have the advantage for the lib user to not have to remember which cells are auto-registered or not so to which `dequeue` calls they have to add `autoRegister: true` vs not add it.

apptekstudios commented:
Thanks for implementing the protocol - I agree that it's definitely a much clearer solution.

I've implemented the AutoRegistering for Headers/Footers and added an example 👍. Hopefully I'll find time to write a quick documentation update tomorrow 🙂

AliSoftware commented:
Ah, bummer.

(I now remember why I'm personally not a fan of doing the auto registering dance in my own projects but prefer explicit registration 😅)

Ok maybe we could indeed register every time before dequeuing (but only for collection views then), but we'd need to clearly document that and the performance trade-off in the documentation to make it crystal clear to users of Reusable then.

Another way would be to keep our own list of registered cell classes per collection view (maybe using ObjC's associated objects?) so we can test before registration. But I wonder if that's not gonna be a similar performance impact anyway…

apptekstudios commented:
I think associated objects is a good way to do it. I went ahead and implemented it 👍

Shouldn't have any performance impact, because all that is done is creating a single object per tableView which then tracks whether the cellType has been registered by us yet.

AliSoftware commented on Sources/View/UICollectionView+Reusable.swift:
```diff
@@ -37,24 +63,70 @@ public extension UICollectionView {
           "Failed to dequeue a cell with identifier \(cellType.reuseIdentifier) matching type \(cellType.self). "
             + "Check that the reuseIdentifier is set properly in your XIB/Storyboard "
-            + "and that you registered the cell beforehand"
+            + "and that you registered the cell beforehand OR enabled autoRegister"
```
(There might be other places with that old mention left maybe, didn't read all the doc comments)

Change to ➡️ "or conformed to `AutoRegistering`"

AliSoftware commented on Sources/View/UICollectionView+Reusable.swift:
```diff
@@ -128,3 +290,44 @@ public extension UICollectionView {
+  private class RegisteredCellTypes {
+    var cellReuseIdentifiers: Set<String>
+    var supplementaryViewReuseIdentifiers: [String: String] //[ElementKind: ReuseIdentifier]
+    init() {
```
Might as well use `=` directly on the properties declaration lines and avoid having to implement `init` there 😉

AliSoftware reviewed (commented):
```
Linting 'InfoViewController.swift' (17/26)
/Users/distiller/project/Example/ReusableDemo iOS/CollectionViewController.swift:19: warning: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
/Users/distiller/project/Example/ReusableDemo iOS/CollectionViewController.swift:23: warning: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace
```

Alex293 commented:
Disclaimer: Did not checked the code so you may be already doing this.

In a project I was doing mostly trhe same thing. For testing if the cell is registered I was trying to dequeue. If it throw -> catch, register then dequeue again

AliSoftware commented:
@Alex293 but doesn't it throw an ObjC `NSException` in those cases? Which can't be caught from Swift (*) and are different from Swift's thrown `Error`s?

---

(*) well, unless if you create an ObjC helper method which catches the `NSException` from ObjC then converts it to an `NSError` returned by reference which is interpreted as a thrown `Error` in Swift… some ObjC pods provide such a helper but I wouldn't want Reusable to depend on another pod just for that — nor to implement that in Reusable itself, requiring to add ObjC to my pod and a bridging header and all that just for that…

AliSoftware commented:
@apptekstudios What do you think about the alternate solution of catching the ObjC exception (instead of going the whole way of using associated objects)?

Something like:

```objc
// ObjCException.h
#import <Foundation/Foundation.h>
@interface ObjC : NSObject
+ (BOOL)catch:(void(^)(void))tryBlock error:(__autoreleasing NSError **)error;
@end

// ObjCException.m
#import "ObjCException.h"
@implementation ObjCException
+ (BOOL)catch:(void(^)(void))tryBlock error:(__autoreleasing NSError **)error {
    @try {
        tryBlock();
        return YES;
    }
    @catch (NSException *exception) {
        *error = [[NSError alloc] initWithDomain:exception.name code:0 userInfo:exception.userInfo];
        return NO;
    }
}
@end
```

Then using it like this from Swift:

```swift
let cell = try? ObjCException.catch {
  return collectionView.dequeue(…)
}
```

apptekstudios commented:
I had a go at catching the objective C exceptions, however couldn't get anything to compile properly with the objective C included in the pod.
https://github.com/apptekstudios/Reusable/tree/objcExceptions

AliSoftware commented:
All right then. Could you just address the nitpickings I made in my latest review? I'll look into merging that next.

apptekstudios committed:
Merge branch 'master' of https://github.com/AliSoftware/Reusable into AliSoftware-master

apptekstudios committed:
Minor Changes

apptekstudios committed:
Update changelog

apptekstudios commented:
Sorry for taking so long to get back to this - I've fixed those minor issues, pulled all changes from your master repo, and updated the changelog.

AliSoftware commented on Sources/View/UITableView+Reusable.swift:
```diff
@@ -47,13 +47,107 @@ public extension UITableView {
+    registerClosure: (() -> Void)?) -> T
+    where T: Reusable {
+      if let registerClosure = registerClosure {
+        if let bareCell = self.dequeueReusableCell(withIdentifier: cellType.reuseIdentifier) as? T {
```
Shouldn't we pass `for: indexPath` in that call here?

Commit:
Merge branch 'master' of https://github.com/AliSoftware/Reusable into AliSoftware-master

Commit:
Merge branch 'AliSoftware-master'

Commit:
Add support to retrieve the topmost view when loading nib content (NibOwnerLoadable)

Commit:
Bump swift version to 5.0
